File: sdkwork-rtc-sdk-typescript/src/provider-package-loader.ts
import { RtcSdkException } from './errors.js';
import type { RtcDriverManager } from './driver-manager.js';
import {
  getRtcProviderPackageByPackageIdentity,
  getRtcProviderPackageByProviderKey,
} from './provider-package-catalog.js';
import {
  registerRtcProviderModule,
  registerRtcProviderModules,
  type RtcProviderModule,
  type RtcProviderModuleDriverOptions,
  type RtcProviderModuleRegistration,
} from './provider-module.js';
import type { RtcProviderPackageCatalogEntry } from './types.js';

export interface RtcProviderPackageLoadRequest {
  providerKey?: string;
  packageIdentity?: string;
}

export interface RtcResolvedProviderPackageLoadTarget {
  providerKey: string;
  packageIdentity: string;
  packageEntry: RtcProviderPackageCatalogEntry;
}

export type RtcProviderModuleNamespace = Record<string, unknown>;

export interface RtcProviderPackageImportFn {
  (
    packageIdentity: string,
    packageEntry: RtcProviderPackageCatalogEntry,
  ): Promise<RtcProviderModuleNamespace>;
}

export interface RtcProviderPackageLoader {
  (
    target: RtcResolvedProviderPackageLoadTarget,
  ): Promise<RtcProviderModuleNamespace>;
}

export interface RtcProviderPackageInstallRequest<TNativeClient = unknown>
  extends RtcProviderPackageLoadRequest {
  options?: RtcProviderModuleDriverOptions<TNativeClient>;
}

function isObjectRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null;
}

function isRtcProviderModule<TNativeClient = unknown>(
  value: unknown,
): value is RtcProviderModule<TNativeClient> {
  return (
    isObjectRecord(value) &&
    typeof value.packageName === 'string' &&
    isObjectRecord(value.metadata) &&
    typeof value.createDriver === 'function'
  );
}

export function resolveRtcProviderPackageLoadTarget(
  request: RtcProviderPackageLoadRequest,
): RtcResolvedProviderPackageLoadTarget {
  const packageEntryByProviderKey =
    typeof request.providerKey === 'string' && request.providerKey.length > 0
      ? getRtcProviderPackageByProviderKey(request.providerKey)
      : undefined;
  const packageEntryByPackageIdentity =
    typeof request.packageIdentity === 'string' && request.packageIdentity.length > 0
      ? getRtcProviderPackageByPackageIdentity(request.packageIdentity)
      : undefined;

  if (
    typeof request.providerKey === 'string' &&
    request.providerKey.length > 0 &&
    typeof request.packageIdentity === 'string' &&
    request.packageIdentity.length > 0
  ) {
    if (
      !packageEntryByProviderKey ||
      !packageEntryByPackageIdentity ||
      packageEntryByProviderKey.providerKey !== packageEntryByPackageIdentity.providerKey
    ) {
      throw new RtcSdkException({
        code: 'provider_package_identity_mismatch',
        message:
          'RTC provider package request drift detected: providerKey and packageIdentity must resolve to the same official package boundary',
        providerKey: request.providerKey,
        pluginId: packageEntryByProviderKey?.pluginId ?? packageEntryByPackageIdentity?.pluginId,
        details: {
          requestedProviderKey: request.providerKey,
          requestedPackageIdentity: request.packageIdentity,
          resolvedProviderKey: packageEntryByProviderKey?.providerKey,
          resolvedPackageIdentity: packageEntryByPackageIdentity?.packageIdentity,
        },
      });
    }
  }

  const packageEntry = packageEntryByProviderKey ?? packageEntryByPackageIdentity;
  if (!packageEntry) {
    throw new RtcSdkException({
      code: 'provider_package_not_found',
      message:
        'RTC provider package resolution requires a known providerKey or packageIdentity from the official package catalog',
      providerKey: request.providerKey,
      details: {
        providerKey: request.providerKey,
        packageIdentity: request.packageIdentity,
      },
    });
  }

  return {
    providerKey: packageEntry.providerKey,
    packageIdentity: packageEntry.packageIdentity,
    packageEntry,
  };
}

export function createRtcProviderPackageLoader(
  importPackage: RtcProviderPackageImportFn = async (packageIdentity) => import(packageIdentity),
): RtcProviderPackageLoader {
  return async (target) => {
    try {
      return await importPackage(target.packageIdentity, target.packageEntry);
    } catch (error) {
      throw new RtcSdkException({
        code: 'provider_package_load_failed',
        message: `Failed to load RTC provider package: ${target.packageIdentity}`,
        providerKey: target.providerKey,
        pluginId: target.packageEntry.pluginId,
        details: {
          packageIdentity: target.packageIdentity,
          moduleSymbol: target.packageEntry.moduleSymbol,
        },
        cause: error,
      });
    }
  };
}

export async function loadRtcProviderModule<TNativeClient = unknown>(
  request: RtcProviderPackageLoadRequest,
  loader: RtcProviderPackageLoader,
): Promise<RtcProviderModule<TNativeClient>> {
  const target = resolveRtcProviderPackageLoadTarget(request);

  let namespace: RtcProviderModuleNamespace;
  try {
    namespace = await loader(target);
  } catch (error) {
    if (error instanceof RtcSdkException) {
      throw error;
    }

    throw new RtcSdkException({
      code: 'provider_package_load_failed',
      message: `Failed to load RTC provider package: ${target.packageIdentity}`,
      providerKey: target.providerKey,
      pluginId: target.packageEntry.pluginId,
      details: {
        packageIdentity: target.packageIdentity,
        moduleSymbol: target.packageEntry.moduleSymbol,
      },
      cause: error,
    });
  }

  if (!isObjectRecord(namespace)) {
    throw new RtcSdkException({
      code: 'provider_module_export_missing',
      message: `RTC provider package did not return a module namespace object: ${target.packageIdentity}`,
      providerKey: target.providerKey,
      pluginId: target.packageEntry.pluginId,
      details: {
        packageIdentity: target.packageIdentity,
        moduleSymbol: target.packageEntry.moduleSymbol,
      },
    });
  }

  const providerModule = namespace[target.packageEntry.moduleSymbol];
  if (!isRtcProviderModule<TNativeClient>(providerModule)) {
    throw new RtcSdkException({
      code: 'provider_module_export_missing',
      message: `RTC provider package is missing the required provider module export: ${target.packageEntry.moduleSymbol}`,
      providerKey: target.providerKey,
      pluginId: target.packageEntry.pluginId,
      details: {
        packageIdentity: target.packageIdentity,
        moduleSymbol: target.packageEntry.moduleSymbol,
        availableExports: Object.keys(namespace),
      },
    });
  }

  return providerModule;
}

export async function installRtcProviderPackage<TNativeClient = unknown>(
  manager: RtcDriverManager,
  request: RtcProviderPackageInstallRequest<TNativeClient>,
  loader: RtcProviderPackageLoader,
): Promise<RtcDriverManager> {
  const providerModule = await loadRtcProviderModule<TNativeClient>(request, loader);
  registerRtcProviderModule(manager, providerModule, request.options);
  return manager;
}

export async function installRtcProviderPackages<TNativeClient = unknown>(
  manager: RtcDriverManager,
  requests: readonly RtcProviderPackageInstallRequest<TNativeClient>[],
  loader: RtcProviderPackageLoader,
): Promise<RtcDriverManager> {
  const registrations: RtcProviderModuleRegistration<TNativeClient>[] = [];

  for (const request of requests) {
    registrations.push({
      providerModule: await loadRtcProviderModule<TNativeClient>(request, loader),
      options: request.options,
    });
  }

  registerRtcProviderModules(manager, registrations);
  return manager;
}

File: sdkwork-rtc-sdk-python/sdkwork_rtc_sdk/provider_package_loader.py
from dataclasses import dataclass
from typing import Any, Callable

from .provider_package_catalog import (
    RtcProviderPackageCatalogEntry,
    get_rtc_provider_package_by_package_identity,
    get_rtc_provider_package_by_provider_key,
)


class RtcProviderPackageLoaderException(Exception):
    def __init__(self, code: str, message: str) -> None:
        super().__init__(message)
        self.code = code


@dataclass(frozen=True)
class RtcProviderPackageLoadRequest:
    providerKey: str | None = None
    packageIdentity: str | None = None


@dataclass(frozen=True)
class RtcResolvedProviderPackageLoadTarget:
    packageEntry: RtcProviderPackageCatalogEntry


RtcProviderModuleNamespace = dict[str, str]
RtcProviderPackageImportFn = Callable[
    [RtcResolvedProviderPackageLoadTarget],
    RtcProviderModuleNamespace,
]
RtcProviderPackageLoader = Callable[
    [RtcProviderPackageLoadRequest],
    RtcProviderModuleNamespace,
]


@dataclass(frozen=True)
class RtcProviderPackageInstallRequest:
    driverManager: Any
    loadRequest: RtcProviderPackageLoadRequest


def resolve_rtc_provider_package_load_target(
    request: RtcProviderPackageLoadRequest,
) -> RtcResolvedProviderPackageLoadTarget:
    package_by_provider_key = (
        get_rtc_provider_package_by_provider_key(request.providerKey)
        if request.providerKey
        else None
    )
    package_by_identity = (
        get_rtc_provider_package_by_package_identity(request.packageIdentity)
        if request.packageIdentity
        else None
    )

    if (
        package_by_provider_key is not None
        and package_by_identity is not None
        and package_by_provider_key.packageIdentity
        != package_by_identity.packageIdentity
    ):
        raise RtcProviderPackageLoaderException(
            "provider_package_identity_mismatch",
            "providerKey and packageIdentity must resolve to the same provider package boundary.",
        )

    resolved_package = package_by_provider_key or package_by_identity
    if resolved_package is None:
        raise RtcProviderPackageLoaderException(
            "provider_package_not_found",
            "No official provider package matches the requested provider boundary.",
        )

    return RtcResolvedProviderPackageLoadTarget(packageEntry=resolved_package)


def create_rtc_provider_package_loader(
    import_package: RtcProviderPackageImportFn,
) -> RtcProviderPackageLoader:
    def _loader(request: RtcProviderPackageLoadRequest) -> RtcProviderModuleNamespace:
        return load_rtc_provider_module(request, import_package)

    return _loader


def load_rtc_provider_module(
    request: RtcProviderPackageLoadRequest,
    import_package: RtcProviderPackageImportFn,
) -> RtcProviderModuleNamespace:
    target = resolve_rtc_provider_package_load_target(request)

    try:
        namespace = import_package(target)
    except RtcProviderPackageLoaderException:
        raise
    except Exception as error:  # pragma: no cover - scaffold-only failure wrapper
        raise RtcProviderPackageLoaderException(
            "provider_package_load_failed",
            f"Reserved provider package loader scaffold could not load {target.packageEntry.packageIdentity}: {error}",
        ) from error

    if not namespace:
        raise RtcProviderPackageLoaderException(
            "provider_module_export_missing",
            "Reserved provider package loader scaffold requires an executable provider module namespace.",
        )

    return namespace


def install_rtc_provider_package(
    request: RtcProviderPackageInstallRequest,
    import_package: RtcProviderPackageImportFn,
) -> None:
    load_rtc_provider_module(request.loadRequest, import_package)
    raise RtcProviderPackageLoaderException(
        "provider_package_load_failed",
        "Reserved provider package installer scaffold cannot register provider modules until a verified runtime bridge lands.",
    )


def install_rtc_provider_packages(
    requests: list[RtcProviderPackageInstallRequest],
    import_package: RtcProviderPackageImportFn,
) -> None:
    for request in requests:
        load_rtc_provider_module(request.loadRequest, import_package)

    if requests:
        raise RtcProviderPackageLoaderException(
            "provider_package_load_failed",
            "Reserved provider package installer scaffold cannot register provider modules until a verified runtime bridge lands.",
        )
